src/Npgsql/MultiHostConnectorPool.cs

```diff
@@ -215,10 +216,17 @@ await TryGet(conn, timeoutPerHost, async, preferredType, IsFallbackOrPreferred,
         }
 
         static NpgsqlException NoSuitableHostsException(IList<Exception> exceptions)
-            => exceptions.Count == 0
-                ? new NpgsqlException("No suitable host was found.")
-                : new("Unable to connect to a suitable host. Check inner exception for more details.",
+        {
+            if (exceptions.Count == 0)
+                return new NpgsqlException("No suitable host was found.");
+
+            var firstException = exceptions[0] as PostgresException;
+            if (firstException is not null && exceptions.All(x => x is PostgresException ex && ex.SqlState == firstException.SqlState))
+                return firstException;
+
+            return new("Unable to connect to a suitable host. Check inner exception for more details.",
                     new AggregateException(exceptions));
+        }
```

Comment:
I like chaining conditional expressions, but up to you. If not, line 228 needs to be indented.

```suggestion
            => exceptions.Count == 0
                ? new NpgsqlException("No suitable host was found.")
                : exceptions[0] is PostgresException firstException &&
                  exceptions.All(x => x is PostgresException ex && ex.SqlState == firstException.SqlState)
                    ? firstException
                    : new NpgsqlException("Unable to connect to a suitable host. Check inner exception for more details.",
                        new AggregateException(exceptions));
```